Too many client cache Invalidation callbacks for a single ICache.clear call

ihsandemir opened the pull request:
Fix for decreasing the cache clear callbacks for the client for a single clear request. Current implementation can cause up to partition count callbacks for a single clear request. One ICache.clear call causes up to partitioncount callbacks to clientcacheinvalidation listener and the same number of near cache clears. This may really be a high number of callbacks and too many cache invalidations if the partition count is a high number. 

The side affect of eliminating the can invalidation event sending during cache store clear needs to be investigated since AbstractCacheRecordStore.clear is being called from a few other places which may have been affected.

Related to issue #8593

jerrinot commented:
won't this completely disable invalidations for member-to-member topologies?

mdogan commented:
What if `cache.clear()` is called from server-side? `CacheClearMessageTask` won't be used at all.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4037/

jerrinot commented:
There are invalidation-related failures as expected.
However I'm surprised the failure are in client tests only?

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4037/

```
com.hazelcast.client.cache.nearcache.ClientNearCacheTest.putToCacheAndClearOrDestroyThenCantGetAnyRecordFromClientNearCacheWithBinaryInMemoryFormat
com.hazelcast.client.cache.nearcache.ClientNearCacheTest.putToCacheAndClearOrDestroyThenCantGetAnyRecordFromClientNearCacheWithObjectInMemoryFormat
```

vbekiaris commented:
Member-side near cache for JCache is not implemented yet.

mdogan commented:
Regardless of member-size near cache, if `cache.clear()` is called on member-side, clients won't get invalidation messages with this change.

vbekiaris commented:
Yes, I agree. It was just meant as a clarification on @jerrinot 's comment about client-only failures.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4042/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4044/

devOpsHazelcast commented:
Test PASSed.

ihsandemir commented:
@mdogan @jerrinot @vbekiaris I implemented the changes for the server side as well. I also added tests to verify that the received event count is really only one as expected for a clear request. I also merged with the new maintenance 3.7.1.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4054/

devOpsHazelcast commented:
Test PASSed.

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java:
```diff
@@ -1455,22 +1455,22 @@ public String getName() {
     }
 
-    protected void onClear() {
+    public void clearAndSendInvalidation() {
```
Maybe this method ought to be part of the `ICacheRecordStore` interface? Please also check this in light of `clear` method usage on EE side.

ihsandemir replied:
I was reluctant to make change to the interface, but as it is not in SPI and inside impl package I will do this.

vbekiaris commented:
I think the PR looks good; please also check for changes required on EE side.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4064/

devOpsHazelcast commented:
Test PASSed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/CachePartitionSegment.java:
```diff
@@ -103,7 +103,8 @@ public void init() {
         synchronized (mutex) {
             for (ICacheRecordStore store : recordStores.values()) {
-                store.clear();
+                AbstractCacheRecordStore recordStore = (AbstractCacheRecordStore) store;
```
do we need this casting any more?

ihsandemir replied:
OK, you are right, changed it now.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/CachePartitionSegment.java:
```diff
@@ -121,7 +122,8 @@ void clearHavingLesserBackupCountThan(int backupCount) {
                 CacheConfig cacheConfig = store.getConfig();
                 if (backupCount > cacheConfig.getTotalBackupCount()) {
-                    store.clear();
+                    AbstractCacheRecordStore recordStore = (AbstractCacheRecordStore) store;
```
same here?

ihsandemir replied:
OK

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/CacheSplitBrainHandler.java:
```diff
@@ -85,6 +86,11 @@ Runnable prepareMergeRunnable() {
+                    // send the cache invalidation event regardless if any actually cleared or not (no need to know how many
+                    // actually cleared)
+                    final CacheService cacheService = nodeEngine.getService(CacheService.SERVICE_NAME);
+                    cacheService.sendInvalidationEvent(cacheName, null, SOURCE_NOT_AVAILABLE);
```
why do we need this?

ihsandemir replied:
I put it so that the old behaviour is not changed. I am not very knowledgable on split brain handling, if it is not needed or we can handle it in a different way, yes I can change it.

ahmetmircik replied:
i see thanks

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheClearMessageTask.java:
```diff
@@ -73,6 +75,11 @@ protected Object reduce(Map<Integer, Object> map) {
+
+        // send the cache invalidation event regardless if any actually cleared or not (no need to know how many actually cleared)
+        final CacheService cacheService = getService(CacheService.SERVICE_NAME);
+        cacheService.sendInvalidationEvent(getDistributedObjectName(), null, SOURCE_NOT_AVAILABLE);
```
seems this should be in `finally` block.

ihsandemir replied:
OK. Changed it.

ahmetmircik commented on hazelcast/src/test/java/com/hazelcast/cache/CacheClearTest.java:
```diff
@@ -165,4 +171,63 @@ public void testClear() {
+                    throws Exception {
+                assertEquals(1, counter.get());
+            }
+        }, 2);
```
2 seconds for `assertTrueEventually` timeout is very short? Does it make sense to use default timeout?

ihsandemir replied:
The test passes with this timeout. The server mostly responds in a shorter time, hence it probably will not fail but yes, we can change to higher timeout. Or we can just wait to see if the test fails in the future due to this reason.

ahmetmircik commented on hazelcast/src/test/java/com/hazelcast/cache/CacheClearTest.java:
```diff
@@ -165,4 +171,63 @@ public void testClear() {
+                    throws Exception {
+                assertTrue(counter.get() <= 1);
+            }
+        }, 3);
```
same here?

devOpsHazelcast commented:
Test PASSed.

ahmetmircik commented:
If it is possible to squeeze commits, it would be better, otherwise LGTM 👍

vbekiaris commented:
👍 (merge once EE counterpart is also ready to merge)

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/4104/

devOpsHazelcast commented:
Test PASSed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractInternalCacheProxy.java:
```diff
@@ -113,6 +114,11 @@ void setCacheManager(HazelcastServerCacheManager cacheManager) {
     }
 
     @Override
+    protected boolean preDestroy() {
```
seems no need to override?

ihsandemir replied:
OK, removed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.
